### ISSUE TITLE
Ability to define restricted record types

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#1811]
I'd like to be able to define a record type such as 

``` javascript
{ 
  color: (string|undefined), 
  size: (number|undefined)
}
```

Currently this is somewhat meaningless because it will match against any object since it will allow unknown properties. So, while I want it to match against any of these:

```
{
  color: "#800"
}
{
  color: "#800",
  size: 5
}
{
  size: 5
}
```

... I don't want it to match these, which it does currently:

```
{
  color: "#800",
  size_typo: 5 
}
```

(Background: this is for use defining React props, to allow templated externs for React that allow the createElement params to be validated).

See also https://github.com/google/closure-compiler/issues/1804#issuecomment-221020830

I've tried using `@record` and `@struct` together although advised this is a grey area, but it did not complain about the additional props. [Debugger](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%250A%2520*%2520%2540record%250A%2520*%2520%2540struct%250A%2520*%252F%250Alet%2520MyType%2520%253D%2520function%28%29%2520%257B%257D%253B%250A%252F**%2520%2540type%2520%257Bstring%257Cundefined%257D%2520*%252F%2520MyType.prototype.color%253B%250A%252F**%2520%2540type%2520%257Bnumber%257Cundefined%257D%2520*%252F%2520MyType.prototype.number%253B%250A%250A%250A%252F**%2520%2540type%2520%257BMyType%257D%2520*%252F%250Alet%2520everything%2520%253D%2520%257B%2520color%253A%2520%2522a%2522%252C%2520size%253A%25205%2520%257D%253B%250A%250A%252F**%2520%2540type%2520%257BMyType%257D%2520*%252F%250Alet%2520missing1%2520%253D%2520%257B%250A%2520%2520%2520%2520color%253A%2520%2522a%2522%250A%257D%253B%250A%250A%252F**%2520%2540type%2520%257BMyType%257D%2520*%252F%250Alet%2520missingAll%2520%253D%2520%257B%250A%257D%253B%250A%250A%252F**%2520%2540type%2520%257BMyType%257D%2520*%252F%250Alet%2520mistyped%2520%253D%2520%257B%250A%2520%2520%2520%2520color%253A%2520%2522a%2522%252C%250A%2520%2520%2520%2520sizes%253A%25205%250A%257D%253B%250A%250A%252F**%2520%2540type%2520%257BMyType%257D%2520*%252F%250Alet%2520mistypedAll%2520%253D%2520%257B%250A%2520%2520%2520%2520colour%253A%2520%2522a%2522%252C%250A%2520%2520%2520%2520sizes%253A%25205%250A%257D%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1).
